crates/providers/src/heart.rs

```diff
@@ -0,0 +1,255 @@
+    // TODO: send a receipt?
+    rx: oneshot::Receiver<()>,
+}
+
+impl PendingTransaction {
+    pub fn tx_hash(&self) -> B256 {
+        self.tx_hash
+    }
+}
+
+impl Future for PendingTransaction {
+    type Output = ();
+
+    fn poll(
+        mut self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Self::Output> {
+        self.rx.poll_unpin(cx).map(|res| res.unwrap())
+    }
+}
+
+/// A handle to the heartbeat task.
+#[derive(Debug, Clone)]
+pub struct HeartbeatHandle {
+    tx: mpsc::Sender<WatchConfig>,
+    latest: watch::Receiver<Block>,
+}
+
+impl HeartbeatHandle {
+    /// Get a new watcher that always sees the latest block.
+    pub fn latest(&self) -> watch::Receiver<Block> {
+        self.latest.clone()
+    }
+}
+
+// TODO: Parameterize with `Network`
+/// A heartbeat task that receives blocks and watches for transactions.
+pub(crate) struct Heartbeat<St> {
+    /// The stream of incoming blocks to watch.
+    stream: futures::stream::Fuse<St>,
+
+    /// Transactions to watch for.
+    unconfirmed: HashMap<B256, WatchConfig>,
+
+    /// Ordered map of transactions waiting for confirmations.
+    waiting_confs: BTreeMap<U256, WatchConfig>,
+
+    /// Ordered map of transactions to reap at a certain time.
+    reap_at: BTreeMap<std::time::Instant, B256>,
+}
+
+impl<St> Heartbeat<St>
+where
+    St: Stream<Item = Block> + Unpin + Send + 'static,
+{
+    /// Create a new heartbeat task.
+    pub fn new(stream: St) -> Self {
+        Self {
+            stream: stream.fuse(),
+            unconfirmed: Default::default(),
+            waiting_confs: Default::default(),
+            reap_at: Default::default(),
+        }
+    }
+}
+
+impl<St> Heartbeat<St> {
+    /// Check if any transactions have enough confirmations to notify.
+    fn check_confirmations(&mut self, latest: &watch::Sender<Block>) {
+        if let Some(current_height) = { latest.borrow().header.number } {
+            let to_keep = self.waiting_confs.split_off(&current_height);
+            let to_notify = std::mem::replace(&mut self.waiting_confs, to_keep);
+
+            for (_, watcher) in to_notify.into_iter() {
+                watcher.notify();
+            }
+        }
+    }
+
+    /// Get the next time to reap a transaction. If no reaps, this is a very
+    /// long time from now (i.e. will not be woken).
+    fn next_reap(&self) -> std::time::Instant {
+        self.reap_at
+            .keys()
+            .next()
+            .copied()
+            .unwrap_or_else(|| std::time::Instant::now() + Duration::from_secs(60_000))
+    }
+
+    /// Reap any timeout
+    fn reap_timeouts(&mut self) {
+        let now = std::time::Instant::now();
+        let to_keep = self.reap_at.split_off(&now);
+        let to_reap = std::mem::replace(&mut self.reap_at, to_keep);
+
+        for tx_hash in to_reap.values() {
+            self.unconfirmed.remove(tx_hash);
+        }
+    }
```

Comment:
Reap is basically "unsubscribe from transactions that have been pending for longer than T"?